### DISCUSSION TITLE
Extending lambda statements with non-lambda expressions

### DISCUSSION THREAD

**Discussion by lonvia** [sqlalchemy/sqlalchemy#10819] [Usage Questions] [answered]
I have a function that creates frequently used SQL statements. To speed this up, it uses lambda statements most of the time. Sometimes the query needs some additional complex WHERE statements that for various reasons cannot be expressed in lambdas. So I've resorted to adding these WHERE statements to the lambda_stmt without lambdas.

A minimal example would look like this:

``` python
import sqlalchemy as sa

t = sa.Table('test', sa.MetaData(),
             sa.Column('id', sa.BigInteger, nullable=False, unique=True),
             sa.Column('s', sa.Text))


def make_query(foo, is_complex):
    sql = sa.lambda_stmt(lambda: sa.select(t).where(t.c.id == foo))

    if is_complex:
        sql = sql.where(t.c.s == 'f') # much more complex in real life

    return sql
```

This is not exactly following the documentation but it seemed to work fine. When `is_complex` is true, then a normal SQL statement is returned and its cache key looks sensible. However, when executing the statement, something goes wrong with the closure variable `foo`.

Lets first create non-complex queries and print the bound parameters:

``` python
sqls = [make_query(x, False) for x in range(2)]
compiled = [s.compile() for s in sqls]

print([c.params for c in compiled])
```

This is fine, the output is:

```
[{'foo_1': 0}, {'foo_1': 1}]
```

Each query has a different value for the bound parameter `foo`.

If I do the same with complex queries:

``` python
sqls = [make_query(x, True) for x in range(2)]
compiled = [s.compile() for s in sqls]

print([c.params for c in compiled])
```

The result is:

```
[{'foo_1': 0, 's_1': 'f'}, {'foo_1': 0, 's_1': 'f'}]
```

The value of  `foo` suddenly is the same for all queries. It looks like it is caching the bind parameters.

I can work around the behaviour in my code, no worries about that. My question is rather: is this a bug in SQLAlchemy worth reporting or am I doing something out-of-spec that was never meant to work?

(Tested on SQLAlchemy 1.4.46 and 2.0.24)

**Comment by zzzeek** [comment, marked as answer]
hi -

kudos for getting into the lambda API and the Core API in general far beyond where most users seem to go.

Short answer I think this is a bug, since it has to do with what happens when you call `.method()` on a lambda, it basically runs the lambda and is supposed to give you a plain construct, but with the whole lambda thing, there's another level or two of caching happening that is all optimized to the usual way lambda statements are used, which is not this.  So we'd assume that's where it goes wrong here and we'd have to change what it does in that case.

so longer answer, context, etc, when I added the transparent caching to SQLAlchemy in version 1.4, which is among other things "most of the guts" of what would also become SQLAlchemy 2.0, to get there I was eventually sprinting really hard to get there, and what we were targeting was something that could completely replace the older "baked query" API, which also uses lambdas.   So I got the lambda thing working well enough that I was able to put it into the ORM loader routines like lazyload/selectinload, replacing the use of the "baked query" system for that part.

But then, it all fell apart.   Too many weird edges / bugs popped out and by something like 1.4.9 or whatever, I had taken the whole thing out of being a fixed part of anything.  fixing issues in the lambda system required extremely meta-meta conceptualizing to do anything, so overall it's too much of a liability.

Really what I want to do is make the cache key generation to be faster without using lambdas (hopefully using cython), then there wont be as much of a benefit to the lambda thing. 

But here, yes, that's a bug, but the thing is if you are inclined to poke around to come up with a fix, that would be great.  or at least some test cases we can add, we can try to figure out how to get this case working.

**Reply by lonvia** [reply]
Thanks for the context. I started looking into query caching and lambdas because I wanted to prove that Python+SQLAlchemy can come close enough to the performance of PHP+handcrafted SQL (see also [this blog post](https://nominatim.org/2023/07/24/performance-revisited.html). It's coming quite close. Kudos for that.). I did quite a few experiments and pretty much the only use case where lambdas do significantly better than query caching alone is with complex queries with very little execution time. So I completely agree that improving cache key generation is the way to go.

I've opened an issue, so there is at least a unit test for the bug. I might poke in the code later to see if I can find a fix but it's not the highest priority.

**Reply by zzzeek** [reply]
ok thanks for that
